src/main/java/org/junit/runners/BlockJUnit4ClassRunner.java

```diff
@@ -463,4 +441,22 @@ private long getTimeout(Test annotation) {
         }
         return annotation.timeout();
     }
+
+    private static final ThreadLocal<RuleContainer> CURRENT_RULE_CONTAINER =
+            new ThreadLocal<RuleContainer>();
```

Comment:
This complexity comes from different methods returning `MethodRule` and `TestRule` instances - both are merged and ordered together.